### PR TITLE
sftp:// support for artifact store

### PR DESCRIPTION
**Pull request by ToonKBC**
Added support for sftp urls as an artifact store:
sftp://server/path
sftp://user:pass@server/path

### CODE REVIEW

**Comment by mateiz** [comment]
FYI it looks like you have a lint warning on Python 2.7. You can follow the Travis link above to see it.

**Comment by mateiz** [comment]
Very cool, thanks! Can you also add a section on these in the MLflow Tracking documentation, the way we added it for the other artifact stores? See https://github.com/mlflow/mlflow/blob/master/CONTRIBUTING.rst#writing-docs for how to build the docs.

File: mlflow/store/sftp_artifact_repo.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            import pysftp
+            import paramiko
+
+            ssh_config = paramiko.SSHConfig()
+            user_config_file = os.path.expanduser("~/.ssh/config")
+            if os.path.exists(user_config_file):
+                with open(user_config_file) as f:
+                    ssh_config.parse(f)
+
+            user_config = ssh_config.lookup(self.config['host'])
```
Comment (smurching) [inline mlflow/store/sftp_artifact_repo.py]: Does this work if `self.config['host']` is `None`, e.g. if someone uses a URI like `sftp:///some/path`? Looks like `lookup` might expect a string (see [Paramiko API docs](http://docs.paramiko.org/en/2.4/api/config.html#paramiko.config.SSHConfig.lookup))

File: mlflow/store/sftp_artifact_repo.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+            if self.config['username'] is None and 'username' in user_config:
+                self.config['username'] = user_config['username']
+
+            if 'identityfile' in user_config:
+                self.config['private_key'] = user_config['identityfile'][0]
```
Comment (smurching) [inline mlflow/store/sftp_artifact_repo.py]: Will `user_config['identityfile']` always be an array? Would be great to confirm if you're able to, the paramiko docs aren't super clear on this

File: mlflow/store/sftp_artifact_repo.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+            self.sftp = pysftp.Connection(**self.config)
+
+        super(SFTPArtifactRepository, self).__init__(artifact_uri)
+
+    def log_artifact(self, local_file, artifact_path=None):
+        artifact_dir = os.path.join(self.path, artifact_path) \
+            if artifact_path else self.path
+        self.sftp.makedirs(artifact_dir)
+        self.sftp.put(local_file, artifact_dir)
```
Comment (smurching) [inline mlflow/store/sftp_artifact_repo.py]: Were you able to confirm if this overwrites `artifact_dir` with `local_file`? Or does it just put `local_file` into `artifact_dir` (the desired behavior)? Looking at the [pysftp](http://pysftp.readthedocs.io/en/release_0.2.9/pysftp.html#pysftp.Connection.put) docs it seems we might overwrite `artifact_dir`

File: docs/source/tracking.rst [hunk 4]
```diff
@@ -263,6 +263,13 @@ in the `GCS documentation <https://google-cloud.readthedocs.io/en/latest/core/au
 Finally, you will need to ``pip install google-cloud-storage`` (on both your client and the server)
 to access Google Cloud Storage; MLflow does not declare a dependency on this package by default.
 
+SFTP server
+~~~~~~~~~~~
+If the specified URI has following format ``sftp://user@host/path/to/directory`` that sftp-server will be used as an artifact store.
```
[the remark this answers is not in the document]
Comment (smurching) [reply docs/source/tracking.rst]: It'd also be good to document that users need to `pip install pysftp` on both the client and server

File: tests/store/test_sftp_artifact_repo.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
+from mock import Mock, MagicMock
+import unittest
+
+import pysftp
+
+from mlflow.store.artifact_repo import ArtifactRepository
+from mlflow.store.sftp_artifact_repo import SFTPArtifactRepository
+
+
+class SFTPArtifactRepositryTest(unittest.TestCase):
```
Comment (smurching) [inline tests/store/test_sftp_artifact_repo.py]: Would it be possible to add tests for `log_artifact` and `log_artifacts`? e.g. we could try logging artifacts to `sftp://<tempdir>` and verify they appear where expected. If such tests require SSH configuration, we could mark them as such via `pytest.mark.requires_ssh` (see [projects CLI tests](https://github.com/mlflow/mlflow/blob/HEAD/tests/projects/test_projects_cli.py#L39))

**Review by smurching — COMMENTED** [review]
Thanks @ToonKBC, this is pretty cool :)

Left a couple of comments, mostly small ones + it'd be nice to add tests for `log_artifact(s)` if possible - thanks!

File: mlflow/store/sftp_artifact_repo.py (hunk 2 above)
Comment (ToonKBC) [reply mlflow/store/sftp_artifact_repo.py]: The ssh_config docs say:
> IdentityFile
> Specifies a file from which the user's RSA or DSA authentication identity is read. The default is ~/.ssh/identity for protocol version 1, and ~/.ssh/id_rsa and ~/.ssh/id_dsa for protocol version 2. Additionally, any identities represented by the authentication agent will be used for authentication.
>The file name may use the tilde syntax to refer to a user's home directory or one of the following escape characters: '%d' (local user's home directory), '%u' (local user name), '%l' (local host name), '%h' (remote host name) or '%r' (remote user name).
>It is possible to have multiple identity files specified in configuration files; all these identities will be tried in sequence.

So one would assume that IdentityFile will be an array. In my case, the returned array has length 1.

File: mlflow/store/sftp_artifact_repo.py (hunk 3 above)
Comment (ToonKBC) [reply mlflow/store/sftp_artifact_repo.py]: I was here wrong indeed. After adding a test, the error was confirmed.

File: mlflow/store/sftp_artifact_repo.py (hunk 1 above)
Comment (ToonKBC) [reply mlflow/store/sftp_artifact_repo.py]: This was indeed an issue. I've added a default case. If `self.config['host'] is None`, `self.config['host'] = 'localhost'`.

**Comment by ToonKBC** [comment]
This was a bit more work than expected. Thanks for the valuable feedback!

**Review by smurching — APPROVED** [review]
Thanks @ToonKBC, this looks good! Looks like there's just a small linter error, after that's fixed let's merge.

Edit - and thank you for the hard work on this :)

File: tests/store/test_sftp_artifact_repo.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    assert len(artifacts) == 2
+    assert artifacts[0].path == dir_name + '/' + file_path
+    assert artifacts[0].is_dir is False
+    assert artifacts[0].file_size == file_size
+    assert artifacts[1].path == dir_name + '/' + subdir_name
+    assert artifacts[1].is_dir is True
+    assert artifacts[1].file_size is None
+
+
+@pytest.mark.requires_ssh
```
Comment (smurching) [inline tests/store/test_sftp_artifact_repo.py]: Actually - I tried running this test locally via `pytest tests/store/test_sftp_artifact_repo.py --requires-ssh` and it failed, do you know what setup is required to run this test?

Comment (ToonKBC) [reply tests/store/test_sftp_artifact_repo.py]: You have to be able to login to ssh without password to localhost. 
If you 'ssh localhost' does it ask for a password? If that's the case, you could add you're pulic key to authorized_keys.

Comment (smurching) [reply tests/store/test_sftp_artifact_repo.py]: Ah got it, thanks for clarifying - using an identityfile + adding the key to authorized_keys worked for me. Would you be able to add a comment stating that passwordless SSH access to localhost is required to run the test? I think at some point we'll add a nightly test runner that runs the tests requiring SSH, so it'd be helpful to document the requirements.

**Comment by smurching** [comment]
Looks like one of the project tests failed due to network failures while accessing GitHub, retriggering the build now

**Comment by smurching** [comment]
LGTM merging, thanks @ToonKBC!

**Pull request merged**